src/AGCertificateVerifier.cpp

```diff
@@ -0,0 +1,973 @@
+                return AGVerifyResult(AGVerifyResult::EXPIRED, X509_verify_cert_error_string(error));
+            case X509_V_ERR_CERT_NOT_YET_VALID:
+                return AGVerifyResult(AGVerifyResult::NOT_YET_VALID, X509_verify_cert_error_string(error));
+            default:
+                return AGVerifyResult(AGVerifyResult::INVALID_CHAIN, X509_verify_cert_error_string(error));
+        }
+    }
+
+    AGVerifyResult res = verifyWeakHashAlgorithm(ctx);
+
+    X509_STORE_CTX_free(ctx);
+    return res;
+}
+
+/**
+ * Verify that certificate matches HPKP pins to hostname
+ * @param dnsName Hostname
+ * @param certChain Certificate chain
+ * @return Verify result
+ */
+AGVerifyResult AGCertificateVerifier::verifyPins(const std::string &dnsName, STACK_OF(X509) *certChain) {
+    // Don't check pins for local root CA
+    if (!verifyChain(mozillaCaStore, certChain).isOk()) {
+        return AGVerifyResult(AGVerifyResult::OK, "Certificate passed basic checks but issued by non-standard CA, skipping pinning test");
+    }
+
+    std::vector<std::string> parts = AGStringUtils::split(dnsName, ".");
+    std::string hostName;
+    for (std::vector<std::string>::const_reverse_iterator i = parts.rbegin(); i != parts.rend(); i++) {
+        hostName = *i + (!hostName.empty() ? "." + hostName : "");
+        std::map<std::string, AGHPKPInfo>::const_iterator j = dynamicHPKPInfo.find(hostName);
+        if (j != dynamicHPKPInfo.end()) {
+            const AGHPKPInfo &info = j->second;
+            if (hostName == info.hostName
+                    && (info.includeSubDomains || hostName == dnsName)
+                    && !info.expired()) {
+                // We have pins for this host, so check'em
+                if (info.hasPinsInChain(certChain)) {
+                    return AGVerifyResult::OK;
+                } else {
+                    return AGVerifyResult(AGVerifyResult::PINNING_ERROR, "Certificate does not match dynamic public key pin for this host");
+                }
+            }
+        }
+        j = staticHPKPInfo.find(hostName);
+        if (j != staticHPKPInfo.end()) {
+            const AGHPKPInfo &info = j->second;
+            if (hostName == info.hostName
+                && (info.includeSubDomains || hostName == dnsName)
+                && !info.expired()) {
+                // We have pins for this host, so check'em
+                if (info.hasPinsInChain(certChain)) {
+                    return AGVerifyResult(AGVerifyResult::OK);
+                } else {
+                    return AGVerifyResult(AGVerifyResult::PINNING_ERROR, "Certificate does not match static public key pin for this host");
+                }
+            }
+        }
+    }
+    return AGVerifyResult::OK;
+}
+
+/**
+ * Save CRLSets CRX to disk
+ * CRLSets CRX extension id: hfnkpimlhhgieaddgfemjhofmfblmnib
+ * CRLSets CRX download URL: http://clients2.google.com/service/update2/crx?response=redirect&x=id%3Dhfnkpimlhhgieaddgfemjhofmfblmnib%26v=%26uc
```

Comment:
It is just hint to library user, where to get CRL sets CRX file to pass to this library. I will make comment more clear.